### PR TITLE
feat: add slack notification processor

### PR DESCRIPTION
**Pull request by kunickiaj**
Adds a new Slack notification processor for use with the notifications plugin.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206973678791022

### CODE REVIEW

**Comment by kunickiaj** [comment]
cc: @drodil

**Comment by drodil** [comment]
cc: @mareklibra @Rugvip

File: plugins/notifications-backend-module-slack/src/config.d.ts
Comment (drodil) [inline plugins/notifications-backend-module-slack/src/config.d.ts:18]: I think this should be optional. When there's more notification configurations, this will not work. I also suggest moving processor configs under `notifications.processors.{processorName}` like they are in for example catalog configuration

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:201]: Maybe you should cache these with `CacheService` as they hardly ever change. This will otherwise call catalog multiple times for each notification.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: You could also add the fields you need from these entities to minimize the data transfer size

**Review by drodil — COMMENTED** [review]
Looks good 👌 Just a couple minor things to consider

File: plugins/notifications-backend-module-slack/src/config.d.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/config.d.ts:18]: Acknowledged on moving. under notifications.processors. Which part should be optional?
I made this similar to catalog providers where you can provide more than one instance (e.g. multiple items in slack array would result in creation of multiple processor instances).

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: Yeah, I considered that. I also was considering using the bulk get endpoint, but I wanted to make sure that if there are N recipients if 1 of them has an issue we still send N-1 notifications. Will see what I can do.

File: plugins/notifications-backend-module-slack/src/config.d.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/config.d.ts:18]: Hmm. Actually these all should not be optional as if the user installs this module, they must also provide configuration, right?

So probably like this:

```ts
notifications: {
  processors: {
    slack: Array<{}>
  }
}
```

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: You could also do something like:
```ts
const users = catalogClient.getEntities(
      {
        filter: [
          {
            kind: 'User',
            [`spec.profile.email`]: CATALOG_FILTER_EXISTS,
          },
        ],
        fields: [
          'kind',
          'metadata.namespace',
          'metadata.name',
          `spec.profile.email`,
        ],
      },
      { token },
);

await cacheService.set('user_emails', users.items.reduce((acc, user) => {
  return acc[stringifyEntityRef(user)] = user.spec.profile.email;
}, {}), {ttl: 60000});
```

To fetch all users with email and cache them to object where the key is user entity reference and value is the email. This would only need one catalog API call every once in a while, depending on the TTL of cache.

Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: Also if you have all the emails of all users, you can send the broadcast notifications to all users pretty easily. Maybe something to consider?

Comment (freben) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: That would be one very large cache entry in larger orgs though. Not sure it's a good fit for some cache types

Comment (freben) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: Using `getEntitiesByRefs` behind a caching (with TTL) `DataLoader` sounds reasonable to me. Then you get automatic rollup into batches without having to care about it in your code.

That endpoint is very fast and handles missing items fine, it just returns null for those entries.

Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: We have about 10,000 users in our catalog and a much smaller subset would actually receive notifications, so I'm not sure I'd go that route for now. I did make some updates just testing them and will update the PR.

Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]: > That would be one very large cache entry in larger orgs though. Not sure it's a good fit for some cache types

True, depends on cache type. Redis for example works pretty well for small cache items so probably caching the users one-by-one would be better option than the whole array under one key.

Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:188]:
```ts
cacheService.set(`${stringifyEntityRef(user)}_email`, user.spec.profile.email, {ttl: 60000});
```

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:152]: I'm still trying to understand how we can provide the best configurability for notifications in general.

In this case I've chosen not to send user notifications if a group is the target entity but that may not always be what is desired.
It would be nice if the NotificationService allowed plugins to register types of notifications emitted. The notification preferences UI could then show all possible notification types and users could choose which NotificationProcessors (delivery methods) should be enabled for each notification type.

Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:164]: You mean something like this: https://github.com/backstage/backstage/pull/23716 ?

File: plugins/notifications-backend-module-slack/src/config.d.ts
Comment (kunickiaj) [inline plugins/notifications-backend-module-slack/src/config.d.ts:29]: These could be left as optional I think for if you want to configure only one of them (e.g. only group/channel support or only direct message support)

I'm fine with either -- optional or both required.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: Maybe something to consider for future how to support these (some specific channel from config, or fetching all users or something?)

File: plugins/notifications-backend-module-slack/src/config.d.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/config.d.ts:29]: Ok, optional is fine 👍

**Comment by drodil** [comment]
Looks good, great job!

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (ydayagi) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts]: this is a great overhead fetching from local DB. The original code is just one short line: 'getEntityByRef'.
The cache and all the logic you display in the comments should be in the catalog API implementations. it is up to the client implementation or server implementation to maintain pools and caches.

Comment (ydayagi) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts]: Do we expect such a large amount of notifications and do we expect it to be to so many users? because if not then the cache is useless here. The size of an entity is small. small text object.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: Maybe a dedicated broadcast channel? TBH I don't think I have a use case for broadcast slack notifications as they tend to become a nuisance if often. Maybe for the announcements plugin could be useful.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts]: Not in the server. That is not something we're looking to provide.


Same with the client. Why would a general purpose client have a cache layer built in? Sounds like a recipe for confusion and bugs. You want to call the device and expect up to date data.

Comment (freben) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts]: The cache is nor about size, but about maintaining a sane request rate toward upstream services that hold data that very rarely changes.

Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts]: I updated the PR to add caching and limit the fields returned. The volume will really depend on implementations and use cases tbh.

Some of the use cases I've identified for us anyway:

* Announcements -- low volume, e.g. new features, PSAs
* Docs Feedback -- pretty low volume
* Pull Request dashboard -- could be much higher volume and many more users. I think caching could be useful here. We have our own internal plugin for this as our setup doesn't really allow us to use standard tooling or GHE's built in PRs view.

All in all, I'm not terribly concerned about the request volume tot he catalog as it seems to be able to handle quite a bit of traffic.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: I think it'd be fine to add broadcast support in a follow-on, esp if its requested.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:164]: Yes.

That reminds me, my original draft had a configuration allowlist for "origins" so that we could easily control _which_ plugins were allowed to send slack notifications. For example you may want some plugins to send slack notifications, but others to limit themselves to web or other processors.

I removed it because origin is only available in the pre/post process methods and not in processOptions and I felt it would then provide an inconsistent experience.

Do you have any ideas on how we could include this? The whole thing may be moot once the notification settings are implemented, but this felt like a way to achieve at least some basic filtering in the meantime.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: Yeah, let's leave it for now like this and add support later if requested. Dedicated channel could be an option, maybe through configuration how the broadcast messages should be handled (to all users / dedicated channel / ignore)

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:37]: Actually thinking about this, could it be also possible to use this annotation for other entities like systems or components and the slack processor could also check those? The default functionality in notifications for other entities is to send the notification to the owner of the entity. But maybe for example components could have specific channel as well?

Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:37]: Yes exactly...

In our catalog groups have these channels and emails optionally configured, and components optionally have them configured as well which would take precedence over the group if defined.

The group is the most common case by far, and the components that have them defined are more the exception but both are supported in our catalog.

Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:37]: I did some quick polling and I think having a notifications specific slack annotation is the route to go as folks tend to have varying preferences and want it to be configurable.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: I made broadcast config support for email here: https://github.com/backstage/backstage/pull/24322 

Maybe something to consider for this as well. Later is ok too.

**Comment by kunickiaj** [comment]
What are the next steps for getting this merged?

**Comment by drodil** [comment]
> What are the next steps for getting this merged?

A proper review from maintainers, @Rugvip or @mareklibra , can you check this out? Thanks!

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (mareklibra) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:101]: I think support for broadcast can be added later.
just a nitpick (feel free to ignore) for this line of code:
```suggestion
       if (!this.channelWebhookUrl || options.recipients.type !== 'entity') {
```

No difference now but it's safe for later.

**Review by mareklibra — COMMENTED** [review]
It looks good to me, just a small nitpick, feel free to ignore

**Comment by kunickiaj** [comment]
No changes, just a rebase while waiting for review approval.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:33]: if you do the human duration change i suggested in the config, this could be `= { hours: 1 }` which is much clearer

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:51]: could also make it required, just always passing it in (moving the responsibility out), but matter of taste

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:106]: optional little thing: `[options.recipients.entityRef].flat()` is short and sweet! But probably a bit more wasteful

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:112]:
```suggestion
        // skip users as they are sent direct messages, but allow all other entity kinds
```

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:140]: does this actually need some rate limiting? a p-limit somewhere

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (freben) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:199]: this sounds like it would benefit from some batching, put a `DataLoader` wrapper around this method I'd say

**Review by freben — APPROVED** [review]
Approving to unblock but had a couple of comments. Main worry is how it'll work under heavy load; this may need some concurrency handling

**Comment by kunickiaj** [comment]
> Approving to unblock but had a couple of comments. Main worry is how it'll work under heavy load; this may need some concurrency handling

Hmm that's a valid point. I looked into Slack's rate limits and for workflow webhook triggers it's pretty low (10/min) which could really be an issue for DMs since each requires a separate hook.

A dedicated Slack App (requires administrator to install) can do 1 message/sec/channel and a few hundred per workspace/min with bursting allowed.

I'll look into what that will change for the processor. My original hope was to avoid needing Slack Admin to install anything.

File: plugins/notifications-backend-module-slack/api-report.md
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/api-report.md:10]: yeah we can shorten the annotation name. Currently this is implemented using Slack Workflows with a Webhook Trigger. The channel ID is passed to the workflow which is specific to a workspace.

Since we'd have to parse out the workspace name as an additional routing param we might also want to shorten that:

e.g. `slack.com/notifications: my-workspace-name/C123456`

We could then augment the config schema to include workspace name for the processor instance.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:140]: I think using `p-throttle` would be better idea, added it to the email processor already so there'a reference and now two use cases.

File: plugins/notifications-backend-module-slack/package.json
Comment (drodil) [inline plugins/notifications-backend-module-slack/package.json:43]: Need to downgrade this to 4.1.1

**Comment by drodil** [comment]
@freben @Rugvip can we get a review for this? thanks!

**Comment by andrewthauer** [comment]
> Hmm that's a valid point. I looked into Slack's rate limits and for workflow webhook triggers it's pretty low (10/min) which could really be an issue for DMs since each requires a separate hook.
> 
> A dedicated Slack App (requires administrator to install) can do 1 message/sec/channel and a few hundred per workspace/min with bursting allowed.
> 
> I'll look into what that will change for the processor. My original hope was to avoid needing Slack Admin to install anything.

I think using the official `@slack/web-api` SDK would be ideal versus webhooks. I feel like these limitations will be outgrown very quickly. We already have a slack app setup and used in our current Backstage instance for some custom things. Would be nice to leverage that out of the box with this module.

File: plugins/notifications-backend-module-slack/api-report.md
Comment (andrewthauer) [reply plugins/notifications-backend-module-slack/api-report.md]: Just thinking a little down the road, but I can potentially foresee it being useful to send certain notification types to different channels. For example a team/group might want to have dedicated slack channels for different purposes. Internally we do this quite often.

Not exactly sure how this would translate to an annotation though as it would need to be at a minimum a list of key/values. Wondering if later on we could append the `spec` to include some sort of notification preferences. Feels a bit wrong, but not sure where else this would naturally fit.

**Comment by freben** [comment]
I'll give it a look again. Just needs a fresh yarn.lock first

**Comment by drodil** [comment]
> > Hmm that's a valid point. I looked into Slack's rate limits and for workflow webhook triggers it's pretty low (10/min) which could really be an issue for DMs since each requires a separate hook.
> > 
> > A dedicated Slack App (requires administrator to install) can do 1 message/sec/channel and a few hundred per workspace/min with bursting allowed.
> > 
> > I'll look into what that will change for the processor. My original hope was to avoid needing Slack Admin to install anything.
> 
> I think using the official `@slack/web-api` SDK would be ideal versus webhooks. I feel like these limitations will be outgrown very quickly. We already have a slack app setup and used in our current Backstage instance for some custom things. Would be nice to leverage that out of the box with this module.

How about supporting both via configuration? I think it's ok to start with webhooks and go from there

File: plugins/notifications-backend-module-slack/api-report.md
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/api-report.md]: Yeah this is something else I've been thinking about. Generally we have private/public/alerts channels, and i think generally we would just use a single channel to route to, routing to multiple channels based on notification type could get complicated quickly.

**Comment by andrewthauer** [comment]
> How about supporting both via configuration? I think it's ok to start with webhooks and go from there

Makes sense to me. Ship this as is and refactor as needed. Might require some breaking changes potentially, but I assume this would be fine given it's current lifecycle?

**Comment by kunickiaj** [comment]
> > Hmm that's a valid point. I looked into Slack's rate limits and for workflow webhook triggers it's pretty low (10/min) which could really be an issue for DMs since each requires a separate hook.
> > A dedicated Slack App (requires administrator to install) can do 1 message/sec/channel and a few hundred per workspace/min with bursting allowed.
> > I'll look into what that will change for the processor. My original hope was to avoid needing Slack Admin to install anything.
> 
> I think using the official `@slack/web-api` SDK would be ideal versus webhooks. I feel like these limitations will be outgrown very quickly. We already have a slack app setup and used in our current Backstage instance for some custom things. Would be nice to leverage that out of the box with this module.

One of my goals was not to require admin setup, but I agree we may hit limits pretty quickly with a busy instance. Perhaps we can support both options.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:159]: For better readability this triple condition could might be explained via extra variable, for example (or some other name like this):

```
const canSendDirectMessage = !this.directWebhookUrl ||
      options.recipients.type === 'broadcast' ||
      !notification.user;
```

Otherwise I'm as a code reader can only suggest the intention there and takes time to understand the purpose of filtering out the notification.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:218]:
```suggestion
      throw new Error(`Email not found for user: ${entityRef}`);
```

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:87]:
```suggestion
    const throttled = throttle((opts: SlackNotificationOptions) => sendSlackNotification(opts));
```

Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:87]: Screenshot 2024-05-29 at 08 25 09

**Comment by kunickiaj** [comment]
Thanks for the review. I've had to focus on other work for a bit but hope to get back to this soon.

Based on the feedback I'm debating whether to continue merging and add support for Slack Apps later or refactor as a Slack App now.

My original intent was to provide an easy, admin permission-less way to provide Slack notifications but for some orgs the rate limits will likely not be sufficient using the Webhook/Workflow method used here and will require a Slack App installed by an admin.

More sophisticated routing can be performed in a Slack app as well, but again, was originally trying to keep things simple and easy to use.

Does anyone here have feedback on the above?

File: plugins/notifications-backend-module-slack/src/lib/constants.ts
Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/constants.ts:20]: This is redundant, as where to send a notification is configured in the workflow.
Instead of this link, the code has to point to the `channelWebhookUrl`

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (acierto) [inline plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:117]: This looks a bit strange, as the destination can be configured in the workflow itself.
That means that it will be always overridden by this logic.

At least it would make sense to keep it as optional, if for some specific entity there should be a different Slack channel then the one which is specified in the workflow itself.

**Comment by acierto** [comment]
> Thanks for the review. I've had to focus on other work for a bit but hope to get back to this soon.
> 
> Based on the feedback I'm debating whether to continue merging and add support for Slack Apps later or refactor as a Slack App now.
> 
> My original intent was to provide an easy, admin permission-less way to provide Slack notifications but for some orgs the rate limits will likely not be sufficient using the Webhook/Workflow method used here and will require a Slack App installed by an admin.
> 
> More sophisticated routing can be performed in a Slack app as well, but again, was originally trying to keep things simple and easy to use.
> 
> Does anyone here have feedback on the above?

As for me, it would make sense to keep both versions. In some orgs you might face security concerns installing custom Applications. Though would be nice to have some minimal support of throttling messages to not exceed the rate limit for the webhook version of it.

File: plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts
Comment (kunickiaj) [reply plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts:117]: The way its intended to be set up is that the destination is not hardcoded in the workflow, but passed in from the notification processor. In the README I suggest setting up the workflow with a variable for the destination.

**Comment by drodil** [comment]
See also #25068

**Comment by phclark** [comment]
I'd love to help this along however I can. We would love to make use of this functionality and think the approach looks great so far!  I do think that a Slack app approach would be worthwhile, and not too out of the norm as most other integrations (eg, auth, scm, CI/CD) typically similar level permissions.

**Comment by phclark** [comment]
Given that the API and the resulting UX are fairly different between the app and webhook approaches, would it make sense to have separate processor plugins?

**Comment by phclark** [comment]
One thing I noticed, when setting this up in my local repository, was that the Slack workflow's webhook field should be `destination`, not `destinationUserEmail` as shown in the included docs. 

I haven't tried sending to a channel yet, but sending to a user works great. I especially love that you're using the user's email instead of handle, as maintaining a mapping between users and their slack handles was going to be a challenge for us.

**Comment by phclark** [comment]
I did get channel notifications working, and think the documentation could be improved to highlight that a separate Slack workflow needs to be created to handle the channel messages vs DMs. 
Also, it may be worth highlighting that the destination webhook field should be of type `slack user email` for the DM workflow's webhook and `Slack channel ID` for the channel workflow's webhook. Likewise, the annotation on the component must actually be a Channel ID, not a channel name as I tried initially.
Once you get past these gotchas, it works great :)

**Comment by drodil** [comment]
Probably not stale?

**Comment by pjungermann** [comment]
> Thanks for the review. I've had to focus on other work for a bit but hope to get back to this soon.
> 
> Based on the feedback I'm debating whether to continue merging and add support for Slack Apps later or refactor as a Slack App now.
> 
> My original intent was to provide an easy, admin permission-less way to provide Slack notifications but for some orgs the rate limits will likely not be sufficient using the Webhook/Workflow method used here and will require a Slack App installed by an admin.
> 
> More sophisticated routing can be performed in a Slack app as well, but again, was originally trying to keep things simple and easy to use.
> 
> Does anyone here have feedback on the above?

I added Slack messages sent to a Slack channel for some Backstage-internal events to our setup a while ago. For that purpose, I used a Slack App. I've created it using a manifest that made it very easy and potentially repeatable. I've considered sharing it eventually.

Sending messages to a channel is easy as well using channel IDs. After the installation, the usability is straightforward and the configuration effort is super simple.

For the basic version, the app required the bot scope `chat:write` which is not that problematic.

I had a couple of more ideas for further features but didn't get back to it, yet. (E.g., interactive features like asking questions/fetching information.)

Looking forward to the final version of this integration.

**Comment by kunickiaj** [comment]
Hi all, just wanted to say that I've been back at work on this as a proper Slack App integration and hope to have an updated PR up soon. Wrapping up some of the work I've been doing internally and hope to have the relevant bits open sourced.

**Comment by drodil** [comment]
Hi @kunickiaj just wondering what happened to this? :)

**Comment by billabongrob** [comment]
@kunickiaj anything we can do to help this along?  We're very interested in getting this past the finish line so we can meet our users where they are.  Happy to help out however possible.

**Pull request closed without merging**
